File: L8_ImageDownload.js
// DOWNLOAD S2 IMAGERY USING SHP OF AOI
// Load shp or draw geometry (comment or uncomment as follows)
// var aoi = table // for uploaded shapefiles
var aoi = geometry // for drawn geometries

// LINES TO EDIT
// 97 - 98 = change start and end date
// 139 - 141 = change scale, image description, and output folder


// Center map to AOI
Map.centerObject(aoi).addLayer(aoi, {}, 'AOI extent')

// mask L8
function maskL8sr (image){
  // mask for clouds
  var cloudShadowBitMask = 1<<3;
  var cloudsBitMask = 1<<5;
  // get QA pixel band (contains info about clouds)
  var qa = image.select('QA_PIXEL');
  // set flags to zero to indicate clear conditions
  var mask = qa.bitwiseAnd(cloudShadowBitMask).eq(0)
               .and(qa.bitwiseAnd(cloudsBitMask).eq(0));
               
  var saturationMask = image.select('QA_RADSAT').eq(0);
               
  // scaling factors
  var opticalBands = image.select('SR_B.').multiply(0.0000275).add(-0.2); // multiply by scale and add offset
  var thermalBands = image.select('ST_B10').multiply(0.00341802).add(149);
               
  return image.addBands(opticalBands, null, true)
              .addBands(thermalBands, null, true)
              .updateMask(mask)
              .updateMask(saturationMask);
}

// INDICES FUNCTIONS:
// the following lines create functions that calculate each index,
// take an image / select the bands to calculate the index / create a layer of the index results /
// returns the original image with the index as a new band
// NDVI 
var addNDVI = function(image){
  var ndvi = image.normalizedDifference(['nir', 'red']).rename('NDVI');
  return image.addBands(ndvi);
}
// SAVI
var addSAVI = function(image){
  var savi = image.expression(
    '1.5*((NIR - RED)/(NIR+RED+0.5))',
    {'NIR': image.select('nir')
    ,'RED': image.select('red')
    }).rename('SAVI');
    return image.addBands(savi).float();
};
// EVI
var addEVI = function(image){
  var evi = image.expression(
    '2.5 *((NIR-RED)/(NIR+6*RED-7.5*BLUE+1))',
    {'NIR': image.select('nir')
    ,'RED': image.select('red')
    ,'BLUE': image.select('blue')
    }).rename('EVI').float();
    return image.addBands(evi);
};
// GCI 
var addGCI = function(image){
  var gci = image.expression(
    '(NIR) / (GREEN)-1',
    {'NIR': image.select('nir')
    ,'GREEN': image.select('green')
    }).rename('GCI').float();
    return image.addBands(gci);
};
// ARVI
var addARVI = function(image){
  var arvi = image.expression(
    '(NIR-(RED-1*(BLUE-RED)))/(NIR+(RED-1*(BLUE-RED)))',
    {'NIR': image.select('nir')
    ,'RED': image.select('red')
    ,'BLUE': image.select('blue')
    }).rename('ARVI').float();
    return image.addBands(arvi);
};
// VARI
var addVARI = function(image){
  var vari = image.expression(
    '(GREEN-RED)/(GREEN+RED-BLUE)',
    {'GREEN': image.select('green')
    ,'RED': image.select('red')
    ,'BLUE': image.select('blue')
    }).rename('VARI').float();
    return image.addBands(vari);
};

// LOAD IMAGERY FROM LANDSAT 8
// 1. Set start and end date (starting from 2013-03-18)
var start_date = '2021-01-01'
var end_date = '2021-12-31'

// 2. Get Landsat imagery
var imagery = ee.ImageCollection('LANDSAT/LC08/C02/T1_L2')
                .filterDate(start_date, end_date)
                .filter(ee.Filter.lte('CLOUD_COVER', 10)) // filter images with <=10% of cloud cover
                .map(maskL8sr)
                .select(['SR_B2', 'SR_B3', 'SR_B4', 'SR_B5','SR_B6', 'SR_B7', 'ST_B10'],
                        ['blue', 'green', 'red', 'nir', 'swir1', 'swir2', 'surface_temp'])
                .map(addNDVI)
                .map(addSAVI)
                .map(addEVI)
                .map(addGCI)
                .map(addARVI)
                .map(addVARI)
                .filterBounds(aoi);
print(imagery)

// 3. Use reducers
// mean
var mean = imagery.select(['blue', 'green', 'red', 'nir', 'swir1', 'swir2', 'surface_temp'
                          ,'NDVI', 'SAVI', 'EVI', 'GCI', 'ARVI', 'VARI'])
                  .mean()
                  .clip(aoi);
// median
var median = imagery.select(['blue', 'green', 'red', 'nir', 'swir1', 'swir2', 'surface_temp'
                          ,'NDVI', 'SAVI', 'EVI', 'GCI', 'ARVI', 'VARI'])
                  .median()
                  .clip(aoi);
// min
var min = imagery.select(['blue', 'green', 'red', 'nir', 'swir1', 'swir2', 'surface_temp'
                          ,'NDVI', 'SAVI', 'EVI', 'GCI', 'ARVI', 'VARI'])
                  .min()
                  .clip(aoi);
// max
var max = imagery.select(['blue', 'green', 'red', 'nir', 'swir1', 'swir2', 'surface_temp'
                          ,'NDVI', 'SAVI', 'EVI', 'GCI', 'ARVI', 'VARI'])
                  .max()
                  .clip(aoi);
                  
// Download the stack
var scale = '10'
var description = 'L8' // change name as desired
var folder_name = 'AGRON665X' // change name to your output folder
// mean
Export.image.toDrive({
      image: mean,
      description: description + '_mean', // change this part according to your desire
      scale: scale,
      folder: folder_name,
      'crs': 'EPSG:4326', 
      region: aoi,
      maxPixels: 1e13
    });
// median
Export.image.toDrive({
      image: median,
      description: description + '_median', // change this part according to your desire
      scale: scale,
      folder: folder_name,
      'crs': 'EPSG:4326', 
      region: aoi,
      maxPixels: 1e13
    });
// min
Export.image.toDrive({
      image: min,
      description: description + '_min', // change this part according to your desire
      scale: scale,
      folder: folder_name,
      'crs': 'EPSG:4326', 
      region: aoi,
      maxPixels: 1e13
    });
// max
Export.image.toDrive({
      image: max,
      description: description + '_max', // change this part according to your desire
      scale: scale,
      folder: folder_name,
      'crs': 'EPSG:4326', 
      region: aoi,
      maxPixels: 1e13
    });
    
// VISUALIZATION OF RESULTS
// Available bands to visualize :
// ['blue', 'green', 'red', 'nir', 'rededge4', 'watervapor', 'swir1', 'swir2',
// 'NDVI', 'SAVI', 'EVI', 'GCI', 'ARVI', 'VARI']
var selection = 'EVI' // enter the name of the band to visualize
//
var visParam = {
  min: 0.0,
  max: 1.8,
  palette: [
    'FFFFFF', 'CE7E45', 'DF923D', 'F1B555', 'FCD163', '99B718', '74A901',
    '66A000', '529400', '3E8601', '207401', '056201', '004C00', '023B01',
    '012E01', '011D01', '011301'
  ],
};
print(mean, 'mean')
print(median, 'median')
print(min, 'min')
print(max, 'max')
Map.addLayer(mean.select(selection), visParam, 'mean')
Map.addLayer(median.select(selection), visParam, 'median')
Map.addLayer(min.select(selection), visParam, 'min')
Map.addLayer(max.select(selection), visParam, 'max')

File: python/L8_ImageDownload.py
import ee
ee.Authenticate()
ee.Initialize()
import geetools
import os
import geopandas as gpd

'''
DOWNLOAD LANDSAT-8 IMAGES
The following program loads images from landsat 8 between two dates, calculates several indices, and exports the images to a Google Drive folder.

Requires:
- Shapefile of the area of interest
- Start and end date to use as reference
- Google drive folder to save the images

Returns:
4 images as TIFF files including bands and indices as layers:
- 'blue', 'green', 'red', 'nir', 'swir1', 'swir2', 'surface_temp', 'NDVI', 'SAVI', 'EVI', 'GCI', 'ARVI', 'VARI'
- each image represents the mean, median, min, and max values for every layer

Edit in lines: 204 - 207
'''


'''
DEFINED FUNCTIONS
'''
# define geometry
def generate_aoi (file_path):
    gdf = gpd.read_file(file_path)
    coord_syst = gdf.crs

    # transform to epsg 4326
    if coord_syst != 'EPSG:4326':
        gdf = gdf.to_crs('EPSG:4326')
        print('TRANSFORMED TO EPSG:4326')

    # border coordinates
    extent_coord = []
    border = gdf.geometry.boundary[0]
    for point in border.coords:
        lon, lat = point
        extent_coord.append((lon, lat))

    # generate aoi
    aoi = ee.Geometry.Polygon(extent_coord)
    print('GEOMETRY GENERATED!')

    return aoi

# mask L8
def maskL8sr (image):
    qa = image.select('QA_PIXEL') # band with cloud information
    cloudShadowBitMask = 1 << 3
    cloudBitMask = 1 << 5

    # set to 0 for clear conditions
    mask = (qa.bitwiseAnd(cloudShadowBitMask).eq(0).And(qa.bitwiseAnd(cloudBitMask).eq(0)))
    
    # saturation mask
    saturationMask = image.select('QA_RADSAT').eq(0)

    # scaling factor
    opticalBands = image.select('SR_B.').multiply(0.0000275).add(-0.2)
    thermalBands = image.select('ST_B10').multiply(0.00312802).add(146)

    return (image.addBands(opticalBands)
            .addBands(thermalBands)
            .updateMask(mask)
            .updateMask(saturationMask))

# clip images
def clipS2 (image):
    return image.clip(aoi)

# NDVI
def addNDVI (image) :
    ndvi = image.normalizedDifference(['nir', 'red']).rename('NDVI')
    return image.addBands(ndvi)
# SAVI
def addSAVI (image):
    savi = image.expression(
    '1.5*((NIR - RED)/(NIR+RED+0.5))',
    {'NIR' : image.select('nir')
    ,'RED' : image.select('red')
    }).rename('SAVI')
    return image.addBands(savi)
# EVI
def addEVI (image):
    evi = image.expression(
    '2.5 *((NIR-RED)/(NIR+6*RED-7.5*BLUE+1))',
    {'NIR': image.select('nir')
    ,'RED': image.select('red')
    ,'BLUE': image.select('blue')
    }).rename('EVI').float()
    return image.addBands(evi)
# GCI 
def addGCI (image):
    gci = image.expression(
    '(NIR) / (GREEN)-1',
    {'NIR': image.select('nir')
    ,'GREEN': image.select('green')
    }).rename('GCI').float()
    return image.addBands(gci)
# ARVI
def addARVI (image):
    arvi = image.expression(
    '(NIR-(RED-1*(BLUE-RED)))/(NIR+(RED-1*(BLUE-RED)))',
    {'NIR': image.select('nir')
    ,'RED': image.select('red')
    ,'BLUE': image.select('blue')
    }).rename('ARVI').float()
    return image.addBands(arvi)
# VARI
def addVARI (image):
    vari = image.expression(
    '(GREEN-RED)/(GREEN+RED-BLUE)',
    {'GREEN': image.select('green')
    ,'RED': image.select('red')
    ,'BLUE': image.select('blue')
    }).rename('VARI').float()
    return image.addBands(vari)
    
# function for downloading imagery
def getLandsat8 (Gdrive_folder, aoi_path, start_date, end_date):
    # generate aoi geometry
    aoi = generate_aoi(aoi_path)

    # get image collection
    imagery = (ee.ImageCollection('LANDSAT/LC08/C02/T1_L2')
               .filterDate(start_date, end_date)
               .filter(ee.Filter.lte('CLOUD_COVER', 10)) #filter for images w/ <= 10 % cloud coverage
               .map(maskL8sr)
               .select(['SR_B2', 'SR_B3', 'SR_B4', 'SR_B5', 'SR_B6', 'SR_B7', 'ST_B10'],
                       ['blue', 'green', 'red', 'nir', 'swir1', 'swir2', 'surface_temp'])
               .map(addNDVI)
               .map(addSAVI)
               .map(addEVI)
               .map(addGCI)
               .map(addARVI)
               .map(addVARI)
               .filterBounds(aoi)           
               )
    
    # reducers
    mean = (imagery.select(['blue', 'green', 'red', 'nir', 'swir1', 'swir2', 'surface_temp', 'NDVI', 'SAVI', 'EVI', 'GCI', 'ARVI', 'VARI'])
            .mean()
            .clip(aoi)
    )

    median = (imagery.select(['blue', 'green', 'red', 'nir', 'swir1', 'swir2', 'surface_temp', 'NDVI', 'SAVI', 'EVI', 'GCI', 'ARVI', 'VARI'])
                    .median()
                    .clip(aoi)
    )

    min = (imagery.select(['blue', 'green', 'red', 'nir', 'swir1', 'swir2', 'surface_temp', 'NDVI', 'SAVI', 'EVI', 'GCI', 'ARVI', 'VARI'])
                    .min()
                    .clip(aoi)
    )

    max = (imagery.select(['blue', 'green', 'red', 'nir', 'swir1', 'swir2', 'surface_temp', 'NDVI', 'SAVI', 'EVI', 'GCI', 'ARVI', 'VARI'])
                    .max()
                    .clip(aoi)
    )

    # export to drive
    file_names = ['mean', 'median', 'min', 'max']
    img_list = [mean, median, min, max]
    scale = 10
    region = aoi.coordinates().getInfo()

    for i, o in enumerate(img_list):
        collection = ee.ImageCollection((o))

        id = i 
        name = "L8_"+file_names[id]

        task = geetools.batch.Export.imagecollection.toDrive(
            collection = collection,
            folder = Gdrive_folder,
            namePattern = name, 
            region = region,
            scale = scale, 
            verbose = True, 
            maxPixels = int(1e13)
        )

    # print summary
    size = imagery.size().getInfo()
    sat = imagery.get('system:id').getInfo()

    print(f'{size} images used from {sat}. n\Data interval between {start_date} and {end_date}')

'''
LOAD IMAGERY FROM LANDSAT-8
'''
### SET UP ###
# Gdrive_folder : folder created in Google Drive to use with GEE
# aoi_path : path to the shapefile that delimits the area of interest
# start date : date to start gathering images (min = 2013-03-18)
# en date : last day to stop gathering images (max = today)

l8 = getLandsat8(Gdrive_folder= 'AGRON665X',
                  aoi_path=r'C:\Users\artur\OneDrive - Iowa State University\4 Codes\DSM tutorial\tutorial\data\boundaries\farm.shp',
                  start_date='2022-01-01',
                  end_date='2023-06-30') 
